### PR TITLE
Add dual-stack support to vSphere CPI

### PR DESCRIPTION
**Pull request by bmhughes**
#### Pull Request Checklist ####

- [ ] Any new images or tags consumed by charts has been added [here](https://github.com/rancher/image-mirror)
- [ ] Chart version has been incremented (if necessary)
- [x] That helm lint and pack run successfully on the chart.
- [x] Deployment of the chart has been tested and verified that it functions as expected.
- [ ] Changes to scripting or CI config have been tested to the best of your ability

#### Types of Change ####

Cilium >= 1.16.0 fails when deployed as dual stack as the nodes will not have IPv6 addresses, the CPI supports adding both IPv4 and IPv6 addresses to nodes but required an feature toggle env var to be specified.

This PR will allow the `ENABLE_ALPHA_DUAL_STACK` env to be set and enable both v4 and v6 address families.

#### After the PR is merged ####

Once the PR is merged, typically upon a new release, the necessary teams will be notified via Slack hook to perform the RKE2 Charts and RKE2 changes. Any developer working on this issue is not responsible for updating RKE2 Charts or RKE2.

### CODE REVIEW

**Review by brandond — APPROVED** [review]

File: charts/rancher-vsphere-cpi/values.yaml [hunk 1]
```diff
@@ -9,12 +9,15 @@ vCenter:
...
 
-# vSphere Tags used to determine the zone and region of a Kubernetes node. This labels will be propagated to NodeLabels
+  # vSphere Tags used to determine the zone and region of a Kubernetes node. This labels will be propagated to NodeLabels
   labels:
     region: "k8s-region"
     zone: "k8s-zone"
     generate: false
 
+  ipFamily: []
+  nodes: {}
```
Comment (jiaqiluo) [inline charts/rancher-vsphere-cpi/values.yaml:19]: Could you add comments to explain the purpose/usage of those two new options?

**Review by jiaqiluo — CHANGES_REQUESTED** [review]
looks good to me overall, I have only one comment

File: charts/rancher-vsphere-cpi/values.yaml (hunk 1 above)
Comment (brandond) [reply charts/rancher-vsphere-cpi/values.yaml:19]: why would you want to statically configure the node list? I've never seen this as a CPI option.

File: charts/rancher-vsphere-cpi/templates/configmap.yaml [hunk 2]
```diff
@@ -30,4 +34,9 @@ data:
       region: {{ .labels.region | quote }}
       zone: {{ .labels.zone | quote }}
     {{- end }}
+
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/templates/configmap.yaml:37]: nit: unnecessary whitespace

File: charts/rancher-vsphere-cpi/values.yaml (hunk 1 above)
Comment (jiaqiluo) [reply charts/rancher-vsphere-cpi/values.yaml:19]: ^^^ +1 

HI @bmhughes, could you provide more context?

Comment (bmhughes) [reply charts/rancher-vsphere-cpi/values.yaml:19]: Sure!

The [ipFamily](https://github.com/kubernetes/cloud-provider-vsphere/blob/a096bb88b449efd2bd980c8855393f3cf593b330/docs/book/cloud_config.md#virtualcenter) option sets the families (v4 or v6) for which the CPI will return the node address for, by default this is only IPv4. With [this](https://github.com/cilium/cilium/pull/28953) commit to cilium it will crash if dual stack is enabled without the node having an IPv6 address present.

[nodes](https://github.com/kubernetes/cloud-provider-vsphere/blob/a096bb88b449efd2bd980c8855393f3cf593b330/docs/book/cloud_config.md#nodes) allows influencing the selection of addresses for internal/external address via cidr/vsphere network name/exclude cidr. Not tied to this but something I thought useful to add.

File: charts/rancher-vsphere-cpi/templates/configmap.yaml (hunk 2 above)
Comment (bmhughes) [reply charts/rancher-vsphere-cpi/templates/configmap.yaml]: Good catch, gone

File: charts/rancher-vsphere-cpi/values.yaml (hunk 1 above)
Comment (bmhughes) [reply charts/rancher-vsphere-cpi/values.yaml:19]: The env is required because the dual stack CPI feature is behind a [flag](https://github.com/kubernetes/cloud-provider-vsphere/blob/a096bb88b449efd2bd980c8855393f3cf593b330/pkg/cloudprovider/vsphere/cloud.go#L56)

Comment (jiaqiluo) [reply charts/rancher-vsphere-cpi/values.yaml:19]: Similarly, for the Nodes part, instead of defining it as a blob {}, could you divide it into 6 individual options? 
Something like this:
```
nodes:
  internalNetworkSubnetCidr: ""
  externalNetworkSubnetCidr: ""
  ..
  .. 
```

That is to not only match the exiting pattern in this file, but also to make it easier to manage.

**Review by jiaqiluo — CHANGES_REQUESTED** [review]

File: charts/rancher-vsphere-cpi/values.yaml (hunk 1 above)
Comment (bmhughes) [reply charts/rancher-vsphere-cpi/values.yaml:19]: For `ipFamily` the docs are a little misleading at they show the legacy ini config where it was a single string and not the [current yaml config](https://github.com/kubernetes/cloud-provider-vsphere/blob/a096bb88b449efd2bd980c8855393f3cf593b330/pkg/common/config/types_yaml.go#L71) where it is a string array.

I'll make the changes for nodes now.

File: charts/rancher-vsphere-cpi/values.yaml [hunk 3]
```diff
@@ -9,12 +9,21 @@ vCenter:
...
     generate: true
 
-# vSphere Tags used to determine the zone and region of a Kubernetes node. This labels will be propagated to NodeLabels
+  # vSphere Tags used to determine the zone and region of a Kubernetes node. This labels will be propagated to NodeLabels
   labels:
     region: "k8s-region"
     zone: "k8s-zone"
     generate: false
 
+  ipFamily: []
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/values.yaml:18]: needs to be a string as noted. Please also add a comment describing usage.

File: charts/rancher-vsphere-cpi/values.yaml [hunk 4]
```diff
@@ -9,12 +9,21 @@ vCenter:
...
+  nodes:
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/values.yaml:19]: please add a comment describing usage

**Comment by brandond** [comment]
Do any of these values need to be added to questions.yaml?

File: charts/rancher-vsphere-cpi/values.yaml (hunk 3 above)
Comment (bmhughes) [reply charts/rancher-vsphere-cpi/values.yaml:18]: For `ipFamily` the docs are a little misleading in that they show the legacy ini config where it was a single string and not the [current yaml config](https://github.com/kubernetes/cloud-provider-vsphere/blob/a096bb88b449efd2bd980c8855393f3cf593b330/pkg/common/config/types_yaml.go#L71) where it is a string array. I've had to figure most of this out from the source rather than the docs.

Will add a comment.

**Comment by bmhughes** [comment]
The nodes options could be easily added. The family is a little more nuanced but could be done, the chart would just have to set the flag env var automatically if family was set.

Might be a nicer experience than `HelmChartConfig`. 🤷🏻‍♂️

**Comment by brandond** [comment]
Rancher already generates a HelmChartConfig based on the values prompted for via questions.yaml. If you put things in there then people can edit them in the UI instead of having to edit the cluster YAML.

**Comment by bmhughes** [comment]
Sounds a better idea, I'll make those changes in the morning. Calling it a night.

**Comment by bmhughes** [comment]
Questions added (I think correctly, docs are a little light). I've tested on a new cluster deploy and works as intended.

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 5]
```diff
@@ -60,3 +60,54 @@ questions:
...
+    default: []
+    required: false
+    group: Configuration
+    options:
+      - ipv4
+      - ipv6
+
+  - variable: vCenter.nodesEnable
+    label: Node address selection filters
+    description: "Define the way that the Node IPs are selected from the addresses assigned to the Node in kube-api"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:77]:
```suggestion
    description: "Define the way that IPs addresses are assigned to the Kubernetes node"
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 6]
```diff
@@ -60,3 +60,54 @@ questions:
         description: vSphere tag which will used to define availability zones
         type: string
         group: Configuration
+
+  - variable: vCenter.ipFamily
+    label: Node Address IP Family
+    description: "IP Family enables the ability to support IPv4 or IPv6. Supported values are: ipv4 - IPv4 addresses only (Default) and ipv6 - IPv6 addresses only"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:66]: The fact that this takes a list makes me suspect that you can actually specify multiple families, and the order sets the preference for the cluster. Is that correct? If so, the description here is incorrect, and should be revised to make it clear that you can pick multiple options, and the order matters.

If this is not correct, and you cannot specify multiple, then how do you actually enable dual-stack clusters?

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 7]
```diff
@@ -60,3 +60,54 @@ questions:
...
+    description: "Define the way that the Node IPs are selected from the addresses assigned to the Node in kube-api"
+    type: boolean
+    default: false
+    required: false
+    group: Configuration
+    show_subquestion_if: true
+    subquestions:
+      - variable: vCenter.nodes.internalNetworkSubnetCidr
+        label: Internal Network CIDR
+        description: "The vSphere cloud provider will select the first address that falls within the provided subnet and assign that value to the Internal network for the node"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:86]:
```suggestion
        description: "The vSphere cloud provider will select the first IP that falls within the provided subnet and assign that value to the Internal address for the node"
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 8]
```diff
@@ -60,3 +60,54 @@ questions:
...
+        type: string
+        group: Configuration
+      - variable: vCenter.nodes.externalNetworkSubnetCidr
+        label: External Network CIDR
+        description: "The vSphere cloud provider will select the first address that falls within the provided subnet and assign that value to the External network for the node"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:91]:
```suggestion
        description: "The vSphere cloud provider will select the first IP that falls within the provided subnet and assign that value to the External address for the node"
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 9]
```diff
@@ -60,3 +60,54 @@ questions:
...
+        type: string
+        group: Configuration
+      - variable: vCenter.nodes.internalVmNetworkName
+        label: Internal VM Network Name
+        description: "The vSphere cloud provider will select the first address found in the VM network matching the provided name and assign that value to the Internal network for the node"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:96]:
```suggestion
        description: "The vSphere cloud provider will select the first IP found in the VM network matching the provided name and assign that value to the Internal address for the node"
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 10]
```diff
@@ -60,3 +60,54 @@ questions:
...
+        type: string
+        group: Configuration
+      - variable: vCenter.nodes.externalVmNetworkName
+        label: External VM Network Name
+        description: "The vSphere cloud provider will select the first address found in the VM network matching the provided name and assign that value to the External network for the node."
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:101]:
```suggestion
        description: "The vSphere cloud provider will select the first IP found in the VM network matching the provided name and assign that value to the External address for the node."
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 11]
```diff
@@ -60,3 +60,54 @@ questions:
...
+        type: string
+        group: Configuration
+      - variable: vCenter.nodes.excludeInternalNetworkSubnetCidr
+        label: Exclude Internal Network CIDR
+        description: "The vSphere cloud provider will never select addresses for the Internal network that fall within the provided subnet ranges. This configuration has the highest precedence"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:106]:
```suggestion
        description: "The vSphere cloud provider will never select IPs for the Internal address that fall within the provided subnet ranges. This configuration has the highest precedence"
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 12]
```diff
@@ -60,3 +60,54 @@ questions:
...
+        type: string
+        group: Configuration
+      - variable: vCenter.nodes.excludeExternalNetworkSubnetCidr
+        label: Exclude External Network CIDR
+        description: "The vSphere cloud provider will never select addresses for the External network that fall within the provided subnet ranges. This configuration has the highest precedence"
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:111]:
```suggestion
        description: "The vSphere cloud provider will never select IPs for the External address that fall within the provided subnet ranges. This configuration has the highest precedence"
```

**Review by brandond — CHANGES_REQUESTED** [review]
It may be too late to change some if it now... but the Labels and Nodes sections are NOT supposed to be nested under vCenter. This is clear from the docs at https://github.com/kubernetes/cloud-provider-vsphere/blob/master/docs/book/cloud_config.md#nodes where there are four sections shown:
* Global
* vCenter
* Labels
* Nodes

The YAML version should have the same structure, except with these at the top-level keys in the map.

I also suggested some improvements to the upstream value descriptions; the wording in the upstream docs is not great and suffers from some copypasta issues.

File: charts/rancher-vsphere-cpi/questions.yaml (hunk 6 above)
Comment (bmhughes) [reply charts/rancher-vsphere-cpi/questions.yaml:66]: Order doesn't matter but you can pick multiple, I'll reword.

**Comment by bmhughes** [comment]
Agree on the value nesting, I'd just carried on as it was before. I can change to make it match upstream.

File: charts/rancher-vsphere-cpi/questions.yaml (hunk 6 above)
Comment (brandond) [reply charts/rancher-vsphere-cpi/questions.yaml:66]: My experience with other CPIs is that the order does matter. If you want a dual-stack cluster with IPv6 as the primary, you need to configure it appropriately.

Comment (bmhughes) [reply charts/rancher-vsphere-cpi/questions.yaml]: I've tried it out and you're correct the order does effect which address is shown with `kubectl get nodes` and on the rancher cluster management page.

It's unclear what this affects but i've updated the description to make note of this.

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 13]
```diff
@@ -60,3 +60,54 @@ questions:
         description: vSphere tag which will used to define availability zones
         type: string
         group: Configuration
+
+  - variable: vCenter.ipFamily
+    label: Node Address IP Family
+    description: "Select the IP family of the address(es) to be assigned to the Node. The first selected family will be the Primary."
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:66]:
```suggestion
    description: "Select the IP families of the address(es) to be assigned to the Node. The first selected family will be the Primary."
```

File: charts/rancher-vsphere-cpi/questions.yaml [hunk 14]
```diff
@@ -63,7 +63,7 @@ questions:
 
   - variable: vCenter.ipFamily
     label: Node Address IP Family
-    description: "Select the IP family of the address(es) to be assigned to the Node. The first selected family will be the Primary."
+    description: "Select the IP families of the address(es) to be assigned to the Node. The first selected family will be the Primary."
     type: enum
```
Comment (brandond) [inline charts/rancher-vsphere-cpi/questions.yaml:67]: question: does `enum` allow you to pick multiple?

Comment (bmhughes) [reply charts/rancher-vsphere-cpi/questions.yaml:67]: Good question I'm not sure, nor how to test it easily. The other option would appear to be list of strings but the documentation is very light here. 

https://github.com/rancher/questions-yml-examples

I've used that as a reference but even that's archived.

Comment (bmhughes) [reply charts/rancher-vsphere-cpi/questions.yaml]: So no `enum` does not allow multiple choices, `listofstring` doesn't seem to work correctly either. I've had to resort to a comma delimited string and then split to a list in helm. Could be nicer but I don't see any other option here.

**Comment by bmhughes** [comment]
@jiaqiluo
@brandond 

Is there anything else required for this to get merged?

**Review by brandond — APPROVED** [review]

**Review by jiaqiluo — APPROVED** [review]

**Comment by brandond** [comment]
We should have required a bump to the chart version before merging this.
